framework/src/org/checkerframework/framework/type/GenericAnnotatedTypeFactory.java

```diff
@@ -896,31 +910,63 @@ protected String dotOutputFileName(UnderlyingAST ast) {
         return null;
     }
 
+    /**
+     * Returns the type of the left-hand side of an assignment without
+     * applying local variable defaults to type variables.
+     *
+     * The type variables that are types of local variables are defaulted to
+     * top so that they can be refined by dataflow.  When these types are used
+     * as context during type argument inference, this default is too conservative.
+     * So this method is used instead of
+     * {@link GenericAnnotatedTypeFactory#getAnnotatedTypeLhs(Tree)}.
+     *
+     * @param lhsTree left-hand side of an assignment
+     * @return AnnotatedTypeMirror of {@code lhsTree}
+     */
+    public AnnotatedTypeMirror getAnnotatedTypeLhsNoTypeVarDefault(Tree lhsTree) {
+        boolean old = this.shouldDefaultTypeVarLocals;
+        shouldDefaultTypeVarLocals = false;
+        AnnotatedTypeMirror type = getAnnotatedTypeLhs(lhsTree);
+        this.shouldDefaultTypeVarLocals = old;
+        return type;
+    }
 
     /**
-     * Get the defaulted type of a variable, without considering
-     * flow inference from the initializer expression.
-     * This is needed to determine the type of the assignment context,
-     * which should have the "default" meaning, without flow inference.
-     * TODO: describe and generalize
+     * Returns the type of a left-hand side of an assignment.
+     *
+     * The default implementation returns the type without considering
+     * dataflow type refinement.  Subclass can override this method
+     * and add additional logic for computing the type of a LHS.
+     *
+     * @param lhsTree left-hand side of an assignment
+     * @return AnnotatedTypeMirror of {@code lhsTree}
      */
-    public AnnotatedTypeMirror getDefaultedAnnotatedType(Tree tree) {
+    public AnnotatedTypeMirror getAnnotatedTypeLhs(Tree lhsTree) {
         AnnotatedTypeMirror res = null;
-        if (tree instanceof VariableTree) {
-            res = fromMember(tree);
-            annotateImplicit(tree, res, false);
-        } else if (tree instanceof AssignmentTree) {
-            res = fromExpression(((AssignmentTree) tree).getVariable());
-            annotateImplicit(tree, res, false);
-        } else if (tree instanceof CompoundAssignmentTree) {
-            res = fromExpression(((CompoundAssignmentTree) tree).getVariable());
-            annotateImplicit(tree, res, false);
-        } else if (TreeUtils.isExpressionTree(tree)) {
-            res = fromExpression((ExpressionTree) tree);
-            annotateImplicit(tree, res, false);
-        } else {
-            assert false;
+        boolean oldUseFlow = useFlow;
+        boolean oldShouldCache = shouldCache;
+        useFlow = false;
+        // Don't cache the result because getAnnotatedType(lhsTree) could
+        // be called and would expect a different result.
```

Comment:
This is a bit confusing, because line 957 calls `getAnnotatedType`. Maybe add something like: later calls to getAnnotatedType will expect flow-sensitive type refinements.